pkg/reconciler/taskrun/taskrun_expired_test.go

```diff
@@ -0,0 +1,168 @@
+package taskrun
+
+import (
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline"
+	apispipeline "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	"github.com/tektoncd/pipeline/test"
+	tb "github.com/tektoncd/pipeline/test/builder"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis"
+)
+
+func newTaskRun(completionTime, failedTime apis.VolatileTime, ttl *metav1.Duration) *apispipeline.TaskRun {
+	tr := tb.TaskRun("test-task-run-with-expiration-ttl", "foo",
+		//tb.TaskRunOwnerReference("PipelineRun", "test-pipeline-run-with-annotations",
+		//	tb.OwnerReferenceAPIVersion("tekton.dev/v1alpha1"),
+		//	tb.Controller, tb.BlockOwnerDeletion,
+		//),
+		tb.TaskRunLabel("tekton.dev/pipeline", "test-pipeline"),
+		tb.TaskRunLabel(pipeline.GroupName+pipeline.PipelineTaskLabelKey, "hello-world-1"),
+		tb.TaskRunLabel("tekton.dev/pipelineRun", "test-pipeline-run-with-annotations"),
+		tb.TaskRunAnnotation("PipelineRunAnnotation", "PipelineRunValue"),
+		tb.TaskRunSpec(
+			tb.TaskRunTaskRef("hello-world"),
+			tb.TaskRunServiceAccountName("test-sa"),
+		),
+	)
+
+	if !completionTime.Inner.IsZero() {
+		c := apis.Condition{Type: apis.ConditionSucceeded, Status: v1.ConditionTrue, LastTransitionTime: completionTime}
+		tr.Status.Conditions = append(tr.Status.Conditions, c)
+	}
+
+	if !failedTime.Inner.IsZero() {
+		c := apis.Condition{Type: apis.ConditionSucceeded, Status: v1.ConditionFalse, LastTransitionTime: failedTime}
+		tr.Status.Conditions = append(tr.Status.Conditions, c)
+	}
+
+	if ttl != nil {
+		tr.Spec.ExpirationSecondsTTL = ttl
+	}
+
+	return tr
+}
+
+func durationPointer(n int) *time.Duration {
+	s := time.Duration(n) * time.Second
+	return &s
+}
+
+func TestTimeLeft(t *testing.T) {
+	now := apis.VolatileTime{Inner: metav1.Now()}
+
+	testCases := []struct {
+		name             string
+		completionTime   apis.VolatileTime
+		failedTime       apis.VolatileTime
+		ttl              *metav1.Duration
+		since            *time.Time
+		expectErr        bool
+		expectErrStr     string
+		expectedTimeLeft *time.Duration
+	}{
+		{
+			name:         "Error case: TaskRun unfinished",
```

Comment:
Also suggest splitting this into two test funcs - one for success cases and one for error cases.